data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java

```diff
@@ -65,85 +70,94 @@
 
-    final CircuitBreakerOptions circuitBreakerOptions = createCircuitBreakerOptions(resource);
+    // recordHandlerFactory allows us to create our record handler `ConsumerRecordHandler`  using the context of the
+    // verticle returned.
+    // (vertx, consumer) -> consumerRecordHandler
+    final BiFunction<Vertx, KafkaConsumer<String, CloudEvent>, Handler<KafkaConsumerRecord<String, CloudEvent>>> recordHandlerFactory = (vertx, consumer) -> {
 
-    final var egressConfig = resource.getEgressConfig();
+      final var producer = createProducer(vertx, resource, egress);
+      final var circuitBreakerOptions = createCircuitBreakerOptions(resource);
+      final var egressConfig = resource.getEgressConfig();
 
-    final var egressDestinationSender = createSender(
-      egress.getDestination(),
-      circuitBreakerOptions,
-      egressConfig
-    );
-    final var egressDeadLetterSender =
-      egressConfig == null || egressConfig.getDeadLetter() == null || egressConfig.getDeadLetter().isEmpty()
-        ? NO_DLQ_SENDER
-        : createSender(egressConfig.getDeadLetter(), circuitBreakerOptions, egressConfig);
+      final var egressSubscriberSender = createSender(
+        vertx,
+        egress.getDestination(),
+        circuitBreakerOptions,
+        egressConfig
+      );
 
-    final var consumerOffsetManager = consumerRecordOffsetStrategyFactory
-      .get(consumer, resource, egress);
+      final var egressDeadLetterSender = isDeadLetterSinkAbsent(egressConfig)
+        ? NO_DLQ_SENDER
+        : createSender(vertx, egressConfig.getDeadLetter(), circuitBreakerOptions, egressConfig);
+
+      return new ConsumerRecordHandler<>(
+        egressSubscriberSender,
+        egress.hasFilter() ? new AttributesFilter(egress.getFilter().getAttributesMap()) : Filter.noop(),
+        this.consumerRecordOffsetStrategyFactory.get(consumer, resource, egress),
+        new HttpSinkResponseHandler(vertx, resource.getTopics(0), producer),
+        egressDeadLetterSender
+      );
+    };
+
+    return new ConsumerVerticle<>(consumerFactory, new HashSet<>(resource.getTopicsList()), recordHandlerFactory);
+  }
 
-    final var sinkResponseHandler = new HttpSinkResponseHandler(vertx, resource.getTopics(0), producer);
+  protected Function<Vertx, KafkaConsumer<String, CloudEvent>> createConsumerFactory(
+    final DataPlaneContract.Resource resource,
+    final DataPlaneContract.Egress egress) {
 
-    final var consumerRecordHandler = new ConsumerRecordHandler<>(
-      egressDestinationSender,
-      (egress.hasFilter()) ? new AttributesFilter(egress.getFilter().getAttributesMap()) : Filter.noop(),
-      consumerOffsetManager,
-      sinkResponseHandler,
-      egressDeadLetterSender
-    );
+    // this.consumerConfigs is a shared object and it acts as a prototype for each consumer instance.
+    final var consumerConfigs = this.consumerConfigs.entrySet()
```

Comment:
Is this  a copy? Can't you just use the new constructor of HashMap that accepts a map?

Reply:
It's not a Map, it's a Properties.

Reply:
and `KafkaClientOptions` takes a `Map<String, Object>`.

Reply:
Since the original `consumerConfigs` is a "template", just make it `Map<String, Object>`

Reply:
Done!